nomad/stream/event_buffer.go

```diff
@@ -0,0 +1,308 @@
+// the head and tail of an atomically updated single-linked list. Atomic
+// accesses are usually to be suspected as premature optimization but this
+// specific design has several important features that significantly simplify a
+// lot of our PubSub machinery.
+//
+// eventBuffer is an adaptation of conuls agent/stream/event eventBuffer but
+// has been updated to be a max length buffer to work for Nomad's usecase.
+//
+// The eventBuffer only tracks the most recent set of published events,
+// up to the max configured size, older events are dropped from the buffer
+// but will only be garbage collected once the slowest reader drops the item.
+// Consumers are notified of new events by closing a channel on the previous head
+// allowing efficient broadcast to many watchers without having to run multiple
+// goroutines or deliver to O(N) separate channels.
+//
+// Because eventBuffer is a linked list with atomically updated pointers, readers don't
+// have to take a lock and can consume at their own pace. Slow readers can eventually
+// append
+//
+// A new buffer is constructed with a sentinel "empty" bufferItem that has a nil
+// Events array. This enables subscribers to start watching for the next update
+// immediately.
+//
+// The zero value eventBuffer is _not_ usable, as it has not been
+// initialized with an empty bufferItem so can not be used to wait for the first
+// published event. Call newEventBuffer to construct a new buffer.
+//
+// Calls to Append or AppendBuffer that mutate the head must be externally
+// synchronized. This allows systems that already serialize writes to append
+// without lock overhead (e.g. a snapshot goroutine appending thousands of
+// events).
+type eventBuffer struct {
+	size *int64
+
+	head atomic.Value
+	tail atomic.Value
+
+	maxSize    int64
+	maxItemTTL time.Duration
+}
+
+// newEventBuffer creates an eventBuffer ready for use.
+func newEventBuffer(size int64, maxItemTTL time.Duration) *eventBuffer {
+	zero := int64(0)
+	b := &eventBuffer{
+		maxSize:    size,
+		size:       &zero,
+		maxItemTTL: maxItemTTL,
+	}
+
+	item := newBufferItem(structs.Events{Index: 0, Events: nil})
+
+	b.head.Store(item)
+	b.tail.Store(item)
+
+	return b
+}
+
+// Append a set of events from one raft operation to the buffer and notify
+// watchers. After calling append, the caller must not make any further
+// mutations to the events as they may have been exposed to subscribers in other
+// goroutines. Append only supports a single concurrent caller and must be
+// externally synchronized with other Append, AppendBuffer or AppendErr calls.
+func (b *eventBuffer) Append(events structs.Events) {
+	b.appendItem(newBufferItem(events))
+}
+
+func (b *eventBuffer) appendItem(item *bufferItem) {
+	// Store the next item to the old tail
+	oldTail := b.Tail()
+	oldTail.link.next.Store(item)
+
+	// Update the tail to the new item
+	b.tail.Store(item)
+
+	// Increment the buffer size
+	size := atomic.AddInt64(b.size, 1)
```

Comment:
I think this is only ever used by externally synchronized methods and therefore doesn't need synchronization? Not a big deal to keep just to avoid races in the future.